packages/core/src/modules/ledger/services/IndyPoolService.ts

```diff
@@ -0,0 +1,154 @@
+      )
+    }
+
+    return this.pools[0]
+  }
+
+  /**
+   * Get the most appropriate pool for the given did. The algorithm is based on the approach as described in this document:
+   * https://docs.google.com/document/d/109C_eMsuZnTnYe2OAd02jAts1vC4axwEKIq7_4dnNVA/edit
+   */
+  public async getPoolForDid(did: string): Promise<{ pool: IndyPool; did: Indy.GetNymResponse }> {
+    const pools = this.pools
+
+    if (pools.length === 0) {
+      throw new LedgerNotConfiguredError(
+        "No indy ledgers configured. Provide at least one pool configuration in the 'indyLedgers' agent configuration"
+      )
+    }
+
+    const { successful, rejected } = await this.getSettledDidResponsesFromPools(did, pools)
+
+    // This means there is a rejected response besides not found.
+    // For now we're really strict, and if one request fails, the whole
+    // request will fail. We may loosen this is the future if we see fit
+    const rejectedOtherThanNotFound = rejected.filter((e) => !(e.reason instanceof LedgerNotFoundError))
+    if (rejectedOtherThanNotFound.length > 0) {
+      throw new LedgerError(
+        `Unknown error retrieving did '${did}' from '${rejectedOtherThanNotFound.length}' of '${pools.length}' ledgers`,
+        { cause: rejectedOtherThanNotFound[0].reason }
+      )
+    }
+
+    // If the successful is empty at this stage, it means all ledgers did not have the did
+    if (successful.length === 0) {
+      throw new LedgerNotFoundError(`Did '${did}' not found on any of the ledgers (total ${this.pools.length}).`)
+    }
+
+    // Split between production and nonProduction ledgers. If there is at least one
+    // successful response from a production ledger, only keep production ledgers
+    // otherwise we only keep the non production ledgers.
+    const production = successful.filter((s) => s.value.pool.config.isProduction)
+    const nonProduction = successful.filter((s) => !s.value.pool.config.isProduction)
+    const remaining = production.length >= 1 ? production : nonProduction
+
+    // If there's only one successful response in this group (production / nonProduction), return it
+    if (remaining.length === 1) {
+      return { pool: remaining[0].value.pool, did: remaining[0].value.did }
+    }
+
+    // If there are multiple successful responses in this group check only one of them is
+    // self-certifying. Return if this is the case
+    const selfCertifying = remaining.filter((response) =>
+      isSelfCertifiedDid(response.value.did.did, response.value.did.verkey)
+    )
+    if (selfCertifying.length === 1) {
+      return { pool: selfCertifying[0].value.pool, did: selfCertifying[0].value.did }
+    }
+
+    // FIXME: ACA-Py doc mentions txnTime, but doc from Stephen mentions that it is not
+    // possible to get when the did was created (it is updated when the did is rotated probably)
+    const sortedByTxnTime = [...remaining].sort(
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

> The sort() method sorts the elements of an array in place

Because JavaScript :)